sdk/src/main/scala/com/horizen/chain/ActiveChain.scala

```diff
@@ -31,88 +33,123 @@ final class ActiveChain private(private val sidechainCache: ElementsChain[Modifi
+    if (mainchainParentHashOpt.isEmpty) throw new IllegalArgumentException ("Parent for mainchain creation block shall be set")
 
-    addToBothStorages(genesisBlockId, genesisBlockInfo, mainchainParentId)
+    // Parent for both first MainchainBlockReferenceData and MainchainHeader is the same for genesis block
+    addToStorages(genesisBlockId, genesisBlockInfo, mainchainParentHashOpt, mainchainParentHashOpt)
   }
 
-  private def setNonGenesisBlock(newBestId: ModifierId, newBestInfo: SidechainBlockInfo, givenMainchainParentId: Option[MainchainBlockReferenceId]): Unit = {
+  private def setNonGenesisBlock(newBestId: ModifierId, newBestInfo: SidechainBlockInfo, givenMainchainParentHashOpt: Option[MainchainHeaderHash]): Unit = {
     // check sidechain correctness
     val parentHeight = heightById(newBestInfo.getParentId).getOrElse(throw new IllegalArgumentException(s"Try to add unconnected sidechain block with id ${newBestId} to an active chain"))
 
-    // check mainchain correctness
-    val actualMainchainParentForNewBlock = getLastMainchainReferenceBeforeHeight(parentHeight).getOrElse(throw new IllegalStateException(s"New best block clear all mainchain references"))
+    // check mainchain headers correctness
+    val actualMainchainParentForNewBlock = getLastMainchainHeaderHashTillHeight(parentHeight).getOrElse(throw new IllegalStateException(s"New best block clear all mainchain headers"))
 
-    val mcReferencesIsEmpty = newBestInfo.mainchainBlockReferenceHashes.isEmpty
-    if (mcReferencesIsEmpty != givenMainchainParentId.isEmpty) {
-      throw new IllegalArgumentException(s"Active chain inconsistency: mainchain references isEmpty are ${mcReferencesIsEmpty} but mainchain parent isEmpty are ${givenMainchainParentId.isEmpty}")
+    val mcReferencesIsEmpty = newBestInfo.mainchainHeaderHashes.isEmpty
+    if (mcReferencesIsEmpty != givenMainchainParentHashOpt.isEmpty) {
+      throw new IllegalArgumentException(s"Active chain inconsistency: mainchain references isEmpty are ${mcReferencesIsEmpty} but mainchain parent isEmpty are ${givenMainchainParentHashOpt.isEmpty}")
     }
 
-    givenMainchainParentId.foreach{givenParentId =>
-      if(givenParentId != actualMainchainParentForNewBlock) {
-        throw new IllegalArgumentException("Try to add inconsistent mainchain block references")
+    givenMainchainParentHashOpt.foreach{givenParentHash =>
+      if(givenParentHash != actualMainchainParentForNewBlock) {
+        throw new IllegalArgumentException("Try to add inconsistent mainchain headers")
       }
     }
 
-    // cut both storages
+    // get actual MainchainReferenceData HeaderHash till parentHeight
+    val actualMainchainReferenceDataParentForNewBlock = getLastMainchainReferenceDataHeaderHashTillHeight(parentHeight)
+      .getOrElse(throw new IllegalStateException(s"New best block clear all mainchain references data header hashes"))
+
```

Comment:
We can also check RefData seq order here:

0. Get current ref data height "dataHeight".
1. Get HeadersHashes from  dataHeight to header height. Append with current headers.
2. Compare hashes with dataHeaderHashes.
Possible use of `chainAfter` on elements chain.